nipype/interfaces/dtitk/registration.py

```diff
@@ -1,394 +1,446 @@
-        return outputs
-
-
-class diffeoSymTensor3DVolInputSpec(CommandLineInputSpec):
-    in_tensor = traits.Str(desc='moving tensor', exists=True, mandatory=False,
-                           position=0, argstr="-in %s")
-    in_xfm = traits.Str(desc='transform to apply', exists=True,
-                        mandatory=False,
-                        position=1, argstr="-trans %s")
-    in_target = traits.Str(desc='', exists=True, mandatory=False, position=2,
-                           argstr="-target %s")
-    out_file = traits.Str(desc='', exists=True, mandatory=False, position=3,
-                          argstr="-out %s", name_source="in_tensor",
-                          name_template="%s_diffeoxfmd.nii.gz")
-
-
-class diffeoSymTensor3DVolOutputSpec(TraitedSpec):
-    out_file = traits.File(desc='cheese', exists=True)
-
-
-class diffeoSymTensor3DVolTask(CommandLineDtitk):
-    """
-    Applies diffeomorphic transform to a tensor volume
-
-                Example
-                -------
-
-                >>> import nipype.interfaces.dtitk as dtitk
-                >>> node = dtitk.diffeoSymTensor3DVolTask()
-                >>> node.inputs.in_tensor = 'diffusion.nii'
-                >>> node.inputs.in_xfm = 'ants_Warp.nii.gz'
-                >>> node.run() # doctest: +SKIP
-                """
-
-    input_spec = diffeoSymTensor3DVolInputSpec
-    output_spec = diffeoSymTensor3DVolOutputSpec
-    _cmd = 'deformationSymTensor3DVolume'
-
-    def _list_outputs(self):
-        outputs = self.output_spec().get()
-        outputs['out_file'] = self.inputs.out_file
+        outputs = self._outputs().get()
+        out_file = self.inputs.out_file
+        if not isdefined(out_file):
+            out_file = self._gen_filename('out_file')
+        outputs['out_file'] = os.path.abspath(out_file)
         return outputs
 
-
-class affSymTensor3DVolInputSpec(CommandLineInputSpec):
-    in_tensor = traits.Str(desc='moving tensor', exists=True, mandatory=False,
-                           position=0, argstr="-in %s")
-    in_xfm = traits.Str(desc='transform to apply', exists=True,
-                        mandatory=False, position=1, argstr="-trans %s")
-    in_target = traits.Str(desc='', exists=True, mandatory=False, position=2,
-                           argstr="-target %s")
-    out_file = traits.Str(desc='', exists=True, mandatory=False, position=3,
-                          argstr="-out %s", name_source="in_tensor",
-                          name_template="%s_affxfmd.nii.gz")
-
-
-class affSymTensor3DVolOutputSpec(TraitedSpec):
-    out_file = traits.File(desc='cheese', exists=True)
-
-
-class affSymTensor3DVolTask(CommandLineDtitk):
+    def _gen_filename(self, name):
+        if name != 'out_file':
+            return
+        return fname_presuffix(os.path.basename(
+                               self.inputs.in_df).split('.')[0],
+                               suffix='_affdf.df.nii')
```

Comment:
So you originally had (a10da5d00df9006ef5bfaf0c80eee8bd5156c273): `name_template="%s_aff.df", keep_extension=True`

I think we can achieve that pattern as follows:

```Python
return fname_presuffix(os.path.basename(self.inputs.in_df),
                       suffix='_aff.df', keep_ext=True)
```

Reply:
In that case, fname_presuffix still keeps the .df, so we end up with an extra extension (`filename.df_aff.df.nii`). I wanted to make it match the other warps ".df.nii" extension with something like "_affdf" prepended so the user knows this warp includes both the affine and diffeomorphic transforms together (`filename_affdf.df.nii` or `filename_affdf.df.nii.gz`). Is the extra extension okay or would it be better to match the other warp files that come from just the diffeo transform?

Reply:
Oh, I see. I misunderstood the situation (and I'm now no longer sure that fixing #2506 will let us do what you want). There are two perspectives here, IMO.

1) File names are semantically meaningful, and as long as we're going to impose a convention within the DTI-TK interfaces, we should respect it.

2) Nipype is often used to bring tools together that have zero respect for each others' conventions, so attempting to manipulate names semantically will fail in some cases. Naming should largely be handled at the `DataSink` level.

I think we can sort of split the difference here:

```Python
path, base, ext = split_filename(self.inputs.in_df)
suffix = '_affdf'
if base.endswith('.df'):
    suffix += '.df'
    base = base[:-3]
return fname_presuffix(base, suffix=suffix, keep_ext=True)
```

This way if you get a file named `x.df.nii`, you'll get `x_affdf.df.nii`, but if you just get `x.nii`, you'll get `x_affdf.nii`.

Did I get your convention right? And does this seem like a reasonable way to achieve that?

Reply:
Yup. Thanks! I think it's a good idea to match the diffeomorphic naming convention.